sources/slack/__init__.py

```diff
@@ -0,0 +1,122 @@
+"""Fetches Slack Conversations, History and logs."""
+
+from typing import Any, Dict, Iterable, Optional
+
+import dlt
+from dlt.common.time import ensure_pendulum_datetime
+from dlt.common.typing import TAnyDateTime, TDataItem
+from dlt.extract.source import DltResource
+from pendulum import DateTime
+
+from .helpers import resource_creator
+from .settings import DEFAULT_START_DATE, MAX_PAGE_SIZE
+
+
+@dlt.source(name="slack")
+def slack_source(
+    access_token: str = dlt.secrets.value,
+    page_size: int = MAX_PAGE_SIZE,
+    start_date: Optional[TAnyDateTime] = DEFAULT_START_DATE,
+    end_date: Optional[TAnyDateTime] = None,
+) -> Iterable[DltResource]:
+    """
+    The source for the Slack pipeline. Available resources are conversations, conversations_history
+    and access_logs.
+
+    Args:
+        access_token: the oauth access_token used to authenticate.
+        page_size: The max number of items to fetch per page. Defaults to 1000.
+        start_date: The start time of the range for which to load. Defaults to January 1st 2000.
+        end_date: The end time of the range for which to load data.
+
+    Returns:
+        Iterable[DltResource]: A list of DltResource objects representing the data resources.
+    """
+
+    def convert_dt(dt: TAnyDateTime, to_ts: bool = False) -> Any:
+        """Converts a datetime to a pendulum datetime or timestamp.
+        Args:
+            dt: The datetime to convert.
+            to_ts: Whether to convert to a timestamp or not.
+        Returns:
+            Any: The converted datetime or timestamp.
+        """
+        if dt is None:
+            return None
+        out_dt = ensure_pendulum_datetime(dt)
+        if to_ts:
+            return out_dt.timestamp()
+        return out_dt
+
+    end_dt: Optional[DateTime] = convert_dt(end_date)
+    start_dt: Optional[DateTime] = convert_dt(start_date)
+
+    @dlt.transformer(
```

Comment:
and i would move this function out of source function. it will make the code more readable